packages/modules/order/src/utils/actions/cancel-item-fulfillment.ts

```diff
@@ -0,0 +1,74 @@
+import { MathBN, MedusaError, isDefined } from "@medusajs/utils"
+import { ChangeActionType } from "../action-key"
+import { OrderChangeProcessing } from "../calculate-order-change"
+
+OrderChangeProcessing.registerActionType(
+  ChangeActionType.CANCEL_ITEM_FULFILLMENT,
+  {
+    operation({ action, currentOrder }) {
+      const existing = currentOrder.items.find(
+        (item) => item.id === action.details.reference_id
+      )!
+
+      existing.detail.fulfilled_quantity ??= 0
+
+      existing.detail.fulfilled_quantity = MathBN.sub(
+        existing.detail.fulfilled_quantity,
+        action.details.quantity
+      )
+    },
+    revert({ action, currentOrder }) {
+      const existing = currentOrder.items.find(
+        (item) => item.id === action.reference_id
+      )!
+
+      existing.detail.fulfilled_quantity = MathBN.add(
+        existing.detail.fulfilled_quantity,
+        action.details.quantity
+      )
+    },
+    validate({ action, currentOrder }) {
+      const refId = action.details?.reference_id
+      if (!isDefined(refId)) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          "Reference ID is required."
+        )
+      }
+
+      const existing = currentOrder.items.find((item) => item.id === refId)
+
+      if (!existing) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          `Reference ID "${refId}" not found.`
```

Comment:
thought(non-blocking): Would it make sense to say `Item ID ${refId} not found`?

Reply:
In this case no much, because the engine is generic.
But we should never see this message working with the module methods.